### PR TITLE
Fixed stack overflow caused by large struct initialization on stack

### PR DESCRIPTION
**Pull request by CircuitCoder**
`FrameAlloc::default()` inside a `lazy_static` initialization clause is called when the variable is accessed for  the first time. The default call is executed on temporary kernel stack, and copy elision is not guaranteed, so it's very possible that a stack overflow may happen.

Replacing the `lazy_static` with a `const fn` can force the compiler to compute the initial value of a variable at compile time. Therefore no initialization is done at runtime.

### CODE REVIEW

File: kernel/Cargo.toml [hunk 1]
```diff
@@ -62,7 +62,7 @@ isomorphic_drivers = { git = "https://github.com/rcore-os/isomorphic_drivers", f
 virtio-drivers = { git = "https://github.com/rcore-os/virtio-drivers", rev = "dfa70e14" }
 lazy_static = { version = "1.4", features = ["spin_no_std"] }
 smoltcp = { git = "https://github.com/rcore-os/smoltcp", rev = "5bd87c7c", default-features = false, features = ["alloc", "log", "ethernet", "proto-ipv4", "proto-igmp", "socket-icmp", "socket-udp", "socket-tcp", "socket-raw"] }
-bitmap-allocator = { git = "https://github.com/rcore-os/bitmap-allocator" }
+bitmap-allocator = { git = "https://github.com/rcore-os/bitmap-allocator", rev="03bd990" }
```
Comment (jiegec) [inline kernel/Cargo.toml:65]: what's the change here

File: kernel/src/sync/mutex.rs [hunk 2]
```diff
@@ -75,11 +77,12 @@ impl<T, S: MutexSupport> Mutex<T, S> {
     ///     drop(lock);
     /// }
     /// ```
-    pub fn new(user_data: T) -> Mutex<T, S> {
+    pub const fn new(user_data: T) -> Mutex<T, S> {
         Mutex {
             lock: AtomicBool::new(false),
             data: UnsafeCell::new(user_data),
-            support: S::new(),
+            support: MaybeUninit::uninit(),
```
Comment (jiegec) [inline kernel/src/sync/mutex.rs:84]: Can we force S::new() to be const fn?

File: kernel/Cargo.toml (hunk 1 above)
Comment (CircuitCoder) [reply kernel/Cargo.toml:65]: A new property `const DEFAULT` was added to the allocator trait. We need that for statically initializing the spinlock

File: kernel/src/sync/mutex.rs (hunk 2 above)
Comment (CircuitCoder) [reply kernel/src/sync/mutex.rs:84]: Currently it's forbidden to have a const fn within a Trait, so I assume there is no way now.

File: kernel/Cargo.toml (hunk 1 above)
Comment (jiegec) [reply kernel/Cargo.toml:65]: why not update Cargo.lock

Comment (CircuitCoder) [reply kernel/Cargo.toml]: Cargo.lock updated

File: kernel/src/sync/mutex.rs [hunk 3]
```diff
@@ -99,7 +102,7 @@ impl<T: ?Sized, S: MutexSupport> Mutex<T, S> {
             let mut try_count = 0;
             // Wait until the lock looks unlocked before retrying
             while self.lock.load(Ordering::Relaxed) {
-                self.support.cpu_relax();
+                unsafe { &*self.support.as_ptr() }.cpu_relax();
```
Comment (jiegec) [inline kernel/src/sync/mutex.rs:105]: get_mut or get_ref?

File: kernel/src/sync/mutex.rs [hunk 4]
```diff
@@ -132,13 +135,38 @@ impl<T: ?Sized, S: MutexSupport> Mutex<T, S> {
...
+                .compare_and_swap(0, 1, Ordering::Acquire)
+                != 0)
+        {
+            // Wait for another thread to initialize
+            while self.support_initialization.load(Ordering::Acquire) == 1 {
+                core::sync::atomic::spin_loop_hint();
+            }
+        } else {
+            // My turn to initialize
+            (unsafe { core::ptr::write(self.support.as_ptr() as *mut _, S::new()) });
```
Comment (jiegec) [inline kernel/src/sync/mutex.rs:165]: .write()?

File: kernel/src/sync/mutex.rs (hunk 3 above)
Comment (CircuitCoder) [reply kernel/src/sync/mutex.rs:105]: It's still an unstable feature, and is completely equivalent to deref and take ref again,

**Pull request merged**